src/Generating/FinishGen.cpp

```diff
@@ -180,6 +180,215 @@ void cFinishGenNetherClumpFoliage::TryPlaceClump(cChunkDesc & a_ChunkDesc, int a
 
 
 ////////////////////////////////////////////////////////////////////////////////
+// cFinishGenClumpTopBlock
+
+void cFinishGenClumpTopBlock::GenFinish(cChunkDesc & a_ChunkDesc)
+{
+	int ChunkX = a_ChunkDesc.GetChunkX();
+	int ChunkZ = a_ChunkDesc.GetChunkZ();
+
+	int NoiseVal = m_Noise.IntNoise2DInt(ChunkX, ChunkZ);
+	EMCSBiome Biome = a_ChunkDesc.GetBiome(cChunkDef::Width / 2, cChunkDef::Width / 2);
+	BiomeInfo info = m_FlowersPerBiome[static_cast<size_t>(Biome)];
+
+	const auto & PossibleBlocks = info.m_Blocks;
+	if (PossibleBlocks.empty())
+	{
+		// No need to go any further. This biome can't generate any blocks.
+		return;
+	}
+
+	int NumClumps = info.m_MaxNumClumpsPerChunk - info.m_MinNumClumpsPerChunk;
+	if (NumClumps == 0)
+	{
+		NumClumps = 1;
+	}
+
+	NumClumps = NoiseVal % NumClumps + info.m_MinNumClumpsPerChunk;
+	int seq = 1;
+	for (int i = 0; i < NumClumps; i++)
+	{
+		int Val1 = m_Noise.IntNoise2DInt(ChunkX * ChunkZ * seq, ChunkZ + ChunkX / seq);
+		int Val2 = m_Noise.IntNoise2DInt(ChunkZ * ChunkX / seq, ChunkZ - ChunkX * seq);
+
+		int PosX = Val1 % (cChunkDef::Width - RANGE_FROM_CENTER * 2) + RANGE_FROM_CENTER;
+		int PosZ = Val2 % (cChunkDef::Width - RANGE_FROM_CENTER * 2) + RANGE_FROM_CENTER;
+
+		int BlockVal = m_Noise.IntNoise2DInt(Val1, Val2);
+
+
+		int TotalWeight = 0;
+		for (const auto & Block : PossibleBlocks)
+		{
+			TotalWeight += Block.m_Weight;
+		}
+
+		// Prevent division by 0
+		TotalWeight = TotalWeight != 0 ? TotalWeight : 1;
+		int Weight = BlockVal % TotalWeight;
+		for (const auto & Block : PossibleBlocks)
+		{
+			Weight -= Block.m_Weight;
+			if (Weight < 0)
+			{
+				TryPlaceFoliageClump(a_ChunkDesc, PosX, PosZ, Block.m_BlockType, Block.m_BlockMeta, Block.m_BlockType == E_BLOCK_BIG_FLOWER);
+				break;
+			}
+		}
+
+		seq++;
+	}
+}
+
+
+
+
+
+void cFinishGenClumpTopBlock::TryPlaceFoliageClump(cChunkDesc & a_ChunkDesc, int a_CenterX, int a_CenterZ, BLOCKTYPE a_BlockType, NIBBLETYPE a_BlockMeta, bool a_IsDoubleTall)
+{
+	int ChunkX = a_ChunkDesc.GetChunkX();
+	int ChunkZ = a_ChunkDesc.GetChunkZ();
+
+	int NumBlocks = m_Noise.IntNoise2DInt(a_CenterX + ChunkX * 16, a_CenterZ + ChunkZ * 16) % (MAX_NUM_FOLIAGE - MIN_NUM_FOLIAGE) + MIN_NUM_FOLIAGE + 1;
+	for (int i = 1; i < NumBlocks; i++)
+	{
+		int x = a_CenterX + ((m_Noise.IntNoise2DInt(i * ChunkX * a_CenterX, i * ChunkZ) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
+		int z = a_CenterZ + ((m_Noise.IntNoise2DInt(i * ChunkZ, i * ChunkX * a_CenterZ) % RANGE_FROM_CENTER * 2) - RANGE_FROM_CENTER);
```

Comment:
Use a single random integer, prefer `+ i` to `* i` (multiplication is much slower on some architectures), divide the random number by a small prime number to make the distribution more flat, then use `(rnd % 256)` and `((rnd / 256) % 256)` as the bases for each direction.